### DISCUSSION TITLE
[V4] Can I render Edge tags dynamically?

### DISCUSSION THREAD

**Discussion by mateus4k** [adonisjs/core#2293] [Help] [answered]
Hello 👋

I want my user to be able to edit email templates on the dashboard and I would like to find a way to do that.

I thought about saving the content of the email in a database table and making it available for editing on the dashboard. For example:
```json
"mail_templates": [
  {
    "id": 1,
    "content": "<h1>Hi, {{user}}</h1>"
  },
  {
    "id": 2,
    "content": "<p>Bye, {{user}}<p>"
  }
]
```
However, how could I go about rendering this template dynamically?

Would the solution be something like the code below?
```js
const edge = require('edge.js');
const data = { user: 'virk' };
const { content } = await TemplateMail.first();
const template = await edge.renderString(content, data);
await Mail.send(template, data);
```
How could I render the content within an Edge layout (using `@layout('layout.mail')`, for example) in this case?

**Comment by thetutlage** [comment, marked as answer]
If it is related to using edge directly, then you can do that using the latest edge release. https://github.com/edge-js/edge/releases/tag/v4.0.2. AdonisJS will have this feature in the next release

**Reply by mateus4k** [reply]
Awesome feature! Will this support `@layout()`?

**Reply by thetutlage** [reply]
Yes, raw strings are 100% compatible with on disk files

**Reply by mateus4k** [reply]
Thanks, Virk! 🚀

**Comment by hbatalhaStch** [comment]
@thetutlage is there a way of doing this in AdonisJS in V5 with a syntax like this:
```ts
await Mail.sendLater((message) => {
  message.htmlView('emails/welcome', {
    templateText: "Welcome  {{ user.fullName }}" // Render params that are inside other params
    user: { fullName: 'Some Name' },
    url: 'https://your-app.com/verification-url',
  })
})
```

**Reply by thetutlage** [reply]
Nope, not really. What you are looking for is rendering a part of text within the template as a template.

What you can do is create a helper function and use that within the template. Something like.

```ts
View.global('renderTemplate', function (contents: string, data: any) {
  return View.renderRaw(contents, data)
})
```

Usage within template.

```edge
{{{ await renderTemplate(templateText, user) }}}
```
